RFC-0018-torch-monitor.md

```diff
@@ -0,0 +1,253 @@
+    COUNT = "count"
+    # SUM computes the sum of the values set within the window.
+    SUM = "sum"
+
+    MAX = "max"
+    MIN = "min"
+
+    # These may not be present in the initial implementation:
+
+    # HISTOGRAM computes summary statistics such as P50, P90, P95, P99
+    HISTOGRAM = "histogram"
+    # STDDEV computes the standard deviation of the values set within
+    # the window.
+    STDDEV = "stddev"
+
+_collectors: Set[StatCollector] = set()
+
+@dataclass
+class Stat:
+    # key is the name of the stat.
+    # Each type of stat should have a globally unique key.
+    key: str
+    # Aggregations is how this stat should be aggregated.
+    aggregations: Set[StatType]
+
+    def add(self, v: float) -> None:
+       for collector in _collectors:
+           collector.handle(self, v)
+
+class StatCollector(ABC):
+    def handle(self, stat: Stat, v: float) -> None:
+        ...
+
+def register_stat_collector(collector: StatCollector) -> None:
+    _collectors.add(collector)
+```
+
+**Example Usage**
+
+```py
+from torch.monitor import Stat, StatType
+
+BATCH_LATENCY = Stat(
+    key="training.batch.latency",
+    aggregations={StatType.MEAN, StatType.HISTOGRAM},
+)
+EPOCH_LATENCY = Stat(
+    key="training.epoch.latency",
+    aggregations={StatType.MEAN, StatType.HISTOGRAM},
+)
+
+def train(...):
+    for i in range(epochs):
+        epoch_start = time.time()
+        for x, y in dataloader:
+            batch_start = time.time()
+            y_pred = model(x)
+            ...
+            BATCH_LATENCY.add(time.time()-batch_start)
+        EPOCH_LATENCY.add(time.time()-epoch_start)
+```
+
+Collectors:
+
+```py
+class AggregatingStatCollector(StatCollector):
+    stats: Set[Stat]
+
+    count: DefaultDict[str, float]
+    sum: DefaultDict[str, float]
+    value: Dict[str, float]
+
+    def handle(self, stat: Stat, v: float) -> None:
+        stats.add(stat)
+
+        if (StatType.MEAN in stat.aggregations
+                or StatType.COUNT in stat.aggregations):
+            self.count[stat.key] += 1
+
+        if (StatType.MEAN in stat.aggregations
+               or StatType.SUM in stat.aggregations):
+           self.sum[stat.key] += v
+
+        if StatType.VALUE in stat.aggregations:
+            self.value[stat.key] = v
+
+        ...
+
+    def report(self) -> Dict[str, float]:
+        out = {}
+        for stat in self.stats:
+            for type in stat.aggregations:
+                if type == StatType.MEAN:
+                    out[stat.key + ".mean"] = (
+                        self.sum[stat.key] / self.count[stat.key]
+                    )
+                ...
+
+    def reset(self) -> None:
+        ...
+
+collector = AggregatingStatCollector()
+register_stat_collector(collector)
+
+# in background thread
+while True:
+   stats = collector.report()
+   collector.reset()
+   for k, v in stats.items():
+      tensorboard_writer.add_scalar(k, v)
+   time.sleep(60)
```

Comment:
For a count based approach the author would specify it on the metric itself? I.e. the author would set that for per epoch data we would log every 1 but for distributed collectives we would log every say 1000. It does add an extra value the author has to configure but seems feasible though potentially susceptible to performance issues depending on the log aggregator

We likely need to support time based aggregation sinks such as Prometheous/Grafana anyways. When logging to time series based systems it doesn't match super well since we may have stale values if the events stop happening. If we default to zero if no items between the Prometheus polls we're likely to get spiky outputs

For the aggregation design, we can export the count as well to help the user handle that behavior

Reply:
I think it's feasible to support both though it does add some more complexity to the aggregation layer. We can have an optional field as part of the stat definition with the # of samples required. 

For conversion to time based loggers we'll likely want to support a "no data" indication so if there's not enough data points we just won't log anything and there will be a gap in the time series -- I believe this should be possible with Prometheus and fb303 though it might be fairly hacky with registering and unregistering series